### PR TITLE
Add SKILL.md parser and skill directory validator

### LINKED ISSUE #3646: SKILL.md parser and validator
## Description

Implement SKILL.md parsing (YAML frontmatter extraction) and skill validation in ToolHive's `pkg/skills/` package. Provides the ability to parse skill metadata from SKILL.md files and validate skills against naming conventions and structural requirements.

## Context

Part of the Skills Lifecycle Management epic. Used by the skill service for install, build, and validate operations.

**Dependencies**: stacklok/toolhive#3645
**Blocks**: Skill service implementation

## Acceptance Criteria

- [ ] `pkg/skills/parser.go` parses SKILL.md YAML frontmatter (name, description, version, allowed-tools, license, requires)
- [ ] Supports `allowed-tools` as space/comma-delimited string or YAML array
- [ ] Parses dependency references from metadata
- [ ] Max 64KB frontmatter, max 100 dependencies
- [ ] `pkg/skills/validator.go` validates skill directories:
  - SKILL.md exists and has valid frontmatter
  - Name matches `^[a-z0-9][a-z0-9-]{0,62}[a-z0-9]$`
  - Required fields present (name, description)
  - No path traversal patterns
  - No symlinks in skill directory
- [ ] Returns structured validation results with errors and warnings
- [ ] All tests pass, SPDX headers present

## References

- stacklok/toolhive#3645
- [Agent Skills Specification](https://agentskills.io/specification)

### PR DESCRIPTION
**Pull request by JAORMX**
## Summary

- Implement SKILL.md frontmatter parsing with YAML extraction, `StringOrSlice` unmarshaler for `allowed-tools`, and dependency parsing from `metadata.skillet.requires`
- Implement skill directory validation: name regex, required fields, field length constraints, name-directory match, symlink detection, path traversal checks
- Add comprehensive table-driven tests (40+ test cases) covering parser and validator edge cases, boundary conditions, and security limits

## Details

Part of the Skills Lifecycle Management epic. Depends on #3645 (stubs/types/interfaces, already merged).

### Parser (`pkg/skills/parser.go` + types in `types.go`)
- `ParseSkillMD()` extracts YAML frontmatter and body from SKILL.md content
- `StringOrSlice` custom YAML unmarshaler supports space-delimited (per Agent Skills spec), comma-delimited, and YAML array formats
- 64KB frontmatter size limit (prevents YAML billion laughs attacks)
- 100 dependency limit for `skillet.requires` metadata

### Validator (`pkg/skills/validator.go`)
- `ValidateSkillDir()` validates a skill directory against the [Agent Skills specification](https://agentskills.io/specification)
- Name validation: `^[a-z0-9][a-z0-9-]{0,62}[a-z0-9]$`, no consecutive hyphens
- Name must match directory name
- Required fields: name, description
- Field constraints: description <= 1024 chars, compatibility <= 500 chars
- Security: symlink detection, path traversal detection
- Warning: SKILL.md > 500 lines

### Types added to `types.go`
- `SkillFrontmatter` - raw YAML frontmatter representation
- `ParseResult` - parsed SKILL.md contents
- `Dependency` - OCI reference dependency
- `StringOrSlice` - flexible YAML unmarshaler

### `ValidationResult` updated in `options.go`
- Added `Warnings []string` field for non-blocking validation warnings

Closes #3646

## Large PR Justification

- Tests take a lot of space here

## Test plan

- [x] `task test` — all unit tests pass (40+ test cases)
- [x] `task lint` — 0 lint issues
- [x] `task license-check` — all SPDX headers present
- [x] Verified against Skillet reference implementation
- [x] Verified against Agent Skills specification

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/skills/types.go [hunk 1]
```diff
@@ -28,6 +34,81 @@ const (
...
+
+// SkillFrontmatter represents the raw YAML frontmatter from a SKILL.md file.
+type SkillFrontmatter struct {
+	Name          string            `yaml:"name"`
+	Description   string            `yaml:"description"`
+	Version       string            `yaml:"version,omitempty"`
+	AllowedTools  StringOrSlice     `yaml:"allowed-tools,omitempty"`
+	License       string            `yaml:"license,omitempty"`
+	Compatibility string            `yaml:"compatibility,omitempty"`
+	Metadata      map[string]string `yaml:"metadata,omitempty"`
```
Comment (aponcedeleonch) [inline pkg/skills/types.go]: ## Inconsistency: `skillet.requires` vs `allowed-tools` list handling

Both `allowed-tools` and `skillet.requires` are conceptually lists of strings, but they're handled very differently:

- `allowed-tools` gets the flexible `StringOrSlice` custom unmarshaler (string or array, multiple delimiters) — great!
- `skillet.requires` is trapped in `map[string]string`, forcing it to be a newline-delimited scalar only. Then `parseRequires()` manually splits on `\n`.

This means if someone writes the natural YAML array form:
```yaml
metadata:
  skillet.requires:
    - ghcr.io/org/skill-a:v1.0.0
    - ghcr.io/org/skill-b:latest
```
it will **silently fail** because `map[string]string` can't hold a `[]string` value.

**Suggestion**: Promote `requires` to a top-level `SkillFrontmatter` field with its own `StringOrSlice` (or a `DependencyList` variant using newline as delimiter). This would:
1. Make it consistent with `allowed-tools`
2. Eliminate the `parseRequires` function and `skilletRequiresKey` constant
3. Support both string and array YAML formats
4. Keep `Metadata` for truly unstructured key-value pairs

Something like:
```go
type SkillFrontmatter struct {
    Name          string            `yaml:"name"`
    Description   string            `yaml:"description"`
    Version       string            `yaml:"version,omitempty"`
    AllowedTools  StringOrSlice     `yaml:"allowed-tools,omitempty"`
    Requires      StringOrSlice     `yaml:"requires,omitempty"`       // promoted
    License       string            `yaml:"license,omitempty"`
    Compatibility string            `yaml:"compatibility,omitempty"`
    Metadata      map[string]string `yaml:"metadata,omitempty"`
}
```

File: pkg/skills/types.go [hunk 2]
```diff
@@ -28,6 +34,81 @@ const (
...
+	case yaml.ScalarNode:
+		str := value.Value
+		if str == "" {
+			*s = nil
+			return nil
+		}
+
+		// Determine delimiter: use comma if present, otherwise space (per spec)
+		var parts []string
+		if strings.Contains(str, ",") {
```
Comment (aponcedeleonch) [inline pkg/skills/types.go:59]: Nit: clean implementation overall. One edge case worth noting: a mixed-delimiter string like `"Read,Glob Grep"` would split on comma (since comma is present), yielding `["Read", "Glob Grep"]` — i.e. `"Glob Grep"` stays as one token. This seems like an acceptable tradeoff given the spec says space-delimited is the canonical format, but might be worth a brief comment documenting the precedence.

File: pkg/skills/parser.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	if !bytes.HasPrefix(content, frontmatterDelimiter) {
+		return nil, nil, ErrInvalidFrontmatter
+	}
+
+	// Skip opening delimiter and optional newline
+	rest := content[len(frontmatterDelimiter):]
+	rest = bytes.TrimPrefix(rest, []byte("\n"))
+
+	endIdx := bytes.Index(rest, frontmatterDelimiter)
```
Comment (aponcedeleonch) [inline pkg/skills/parser.go]: Minor: the size check on line 79 happens *after* `bytes.Index` has already scanned the full `rest` content. This means a 100MB file with no closing `---` would be fully scanned before returning `ErrInvalidFrontmatter` (and never even reach the size check). Consider adding a fast-fail length check before the `bytes.Index` call:

```suggestion
	if len(rest) > MaxFrontmatterSize {
		return nil, nil, fmt.Errorf("%w: content too large for frontmatter parsing",
			ErrInvalidFrontmatter)
	}

	endIdx := bytes.Index(rest, frontmatterDelimiter)
```

File: pkg/skills/validator.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	if strings.Contains(name, "--") {
+		return fmt.Errorf("invalid skill name %q: must not contain consecutive hyphens", name)
+	}
+	return nil
+}
+
+// checkSymlinks walks the directory and checks for symbolic links.
+func checkSymlinks(path string) error {
+	return filepath.Walk(path, func(p string, _ os.FileInfo, err error) error {
```
Comment (aponcedeleonch) [inline pkg/skills/validator.go]: Minor: consider using `filepath.WalkDir` instead of `filepath.Walk` (same for `checkPathTraversal` below). `WalkDir` (Go 1.16+) is more efficient — it doesn't call `Stat` on every entry. Since you're calling `os.Lstat` manually anyway, `WalkDir` would avoid the redundant `Stat` that `Walk` does internally.

Also, both `checkSymlinks` and `checkPathTraversal` walk the same directory tree. They could be combined into a single walk to avoid traversing the tree twice.

File: pkg/skills/types.go (hunk 1 above)
Comment (JAORMX) [reply pkg/skills/types.go]: Great catch, and yeah, the silent failure is a real footgun. The reason `skillet.requires` lives in `metadata` is that it's not part of the [Agent Skills spec](https://agentskills.io/specification) — it's an extension. The spec defines `metadata` as "a map from string keys to string values", so `map[string]string` is by design. Skillet's POC implemented it this way too, but the spec is the source of truth here.

That said, you're right that someone writing the natural YAML array form would get bitten with no warning. I'll add a comment documenting this limitation. For a follow-up, we could add validation that warns if a metadata value looks like it was intended to be an array. WDYT?

File: pkg/skills/types.go (hunk 2 above)
Comment (JAORMX) [reply pkg/skills/types.go:59]: Good call. Will add a comment documenting the precedence: comma wins over space when both are present. The spec says space-delimited is the canonical format; comma and array support are our compatibility extensions.

File: pkg/skills/parser.go (hunk 3 above)
Comment (JAORMX) [reply pkg/skills/parser.go]: Great catch on the ordering. Will add the early size check before `bytes.Index`. Note that we're also somewhat protected since the caller reads via `os.ReadFile` which is bounded by available memory, but the early check is good defense-in-depth. No reason to scan 100MB looking for `---` when we know we'll reject it anyway.

File: pkg/skills/validator.go (hunk 4 above)
Comment (JAORMX) [reply pkg/skills/validator.go]: Good suggestions on both counts. `WalkDir` avoids the redundant `Stat` that `Walk` does internally, and since we're calling `os.Lstat` manually anyway it's a no-brainer. Will also combine both checks into a single walk function. Cleaner and more efficient.

File: pkg/skills/types.go (hunk 2 above)
Comment (aponcedeleonch) [reply pkg/skills/types.go:59]: if space-delimited is the canonical format, wouldn't we want that to have precedence?

Comment (JAORMX) [reply pkg/skills/types.go:59]: Good question! So, basically, comma-first is correct *because* the canonical space-delimited format never contains commas. It naturally falls through to the space path:

- `Read Glob Grep` (canonical) -> no commas found -> splits on space -> `["Read", "Glob", "Grep"]` (correct)
- `Read, Glob, Grep` (compat) -> commas found -> splits on comma -> `["Read", "Glob", "Grep"]` (correct)

If we flipped it and gave space precedence:

- `Read, Glob, Grep` -> splits on space -> `["Read,", "Glob,", "Grep"]` (broken)

The commas would stay stuck to the tool names. Note that tool names themselves never contain commas, so comma presence is a reliable signal that the author intended comma-delimited format.

File: pkg/skills/types.go (hunk 1 above)
Comment (aponcedeleonch) [reply pkg/skills/types.go]: I would add support for writing natural YAML TBH since we're augmenting what the spec supports already. I would also add the warning you're suggesting. That way users know that their skill might not work outside of ToolHive

File: pkg/skills/types.go (hunk 2 above)
Comment (aponcedeleonch) [reply pkg/skills/types.go:59]: ok, sounds good. Maybe we should also at least log a warning that we're doing delimited not supported originally by the spec? Will leave it up to you :)

File: pkg/skills/types.go (hunk 1 above)
Comment (JAORMX) [reply pkg/skills/types.go]: Quick update: we're promoting requires to a top-level frontmatter field (with StringOrSlice, so both YAML arrays and newline-delimited strings work). Also renaming from skillet.requires to toolhive.requires since skillet was the PoC name.
